src/entities/abTrail/ui/AbCanvas.tsx

```diff
@@ -0,0 +1,385 @@
+import { FC, useEffect, useRef, useState } from 'react';
+import { StyleSheet } from 'react-native';
+
+import {
+  Skia,
+  SkiaView,
+  SkPath,
+  TouchInfo,
+  PaintStyle,
+  useTouchHandler,
+  useDrawCallback,
+  SkCanvas,
+} from '@shopify/react-native-skia';
+
+import { Box, BoxProps } from '@app/shared/ui';
+
+import AbShapes from './AbShapes';
+import {
+  DeviceType,
+  LogLine,
+  LogPoint,
+  MessageType,
+  Point,
+  TestNode,
+  TestScreenPayload,
+} from '../lib';
+import { getDistance, transformCoordinates } from '../lib/utils';
+
+const paint = Skia.Paint();
+paint.setColor(Skia.Color('black'));
+paint.setStrokeWidth(1);
+paint.setStyle(PaintStyle.Stroke);
+
+const ErrorLineTimeout = 1000;
+const FlareGreenPointTimeout = 1000;
+
+type ResultLog = {
+  lines: LogLine[];
+  currentIndex: number;
+};
+
+type Props = {
+  testData: TestScreenPayload;
+  deviceType: DeviceType;
+  readonly: boolean;
+  onLogResult: (data: ResultLog) => void;
+  onMessage: (message: MessageType) => void;
+  onComplete: () => void;
+} & BoxProps;
+
+const AbCanvas: FC<Props> = props => {
+  const [canvasData, setCanvasData] = useState<TestScreenPayload | null>(null);
+
+  const [errorPath, setErrorPath] = useState<SkPath | null>(null);
+
+  const [paths, setPaths] = useState<Array<SkPath>>([]);
+
+  const [width, setWidth] = useState<number | null>(null);
+
+  const [flareGreenPointIndex, setFlareGreenPointIndex] = useState<{
+    index: number;
+  } | null>(null);
+
+  const currentPathRef = useRef<SkPath | null>();
+
+  const currentIndexRef = useRef(1);
+
+  const canvasRef = useRef<SkCanvas>();
+
+  const logLines = useRef<LogLine[]>([]).current;
+
+  const { testData, deviceType, onLogResult, onMessage, onComplete, readonly } =
+    props;
+
+  useEffect(() => {
+    if (!testData || !width) {
+      return;
+    }
+    const transformed = transformCoordinates(testData, width);
```

Comment:
Please use useMemo here instead of useEffect,
```
const canvasData = useMemo(
  () => width ? transformCoordinates(testData, width) : null,
  [width, testData]
);
```